File: algorithms/sorting/merge-sort.js
/*
Problem: Given an array of integers, sort the array
 */

function mergeSort(arr) {
  // base case
  if (arr.length < 2) return arr;

  const mid = Math.floor(arr.length / 2);
  const leftArray = arr.slice(0, mid);
  const rightArray = arr.slice(mid);

  return merge(mergeSort(leftArray), mergeSort(rightArray));
}

function merge(leftArray, rightArray) {
  const sortedArray = [];
  while (leftArray.length && rightArray.length) {
    if (leftArray[0] <= rightArray[0]) {
      // use shift since element should be removed from left array
      sortedArray.push(leftArray.shift());
    } else {
      sortedArray.push(rightArray.shift());
    }
  }
  return [...sortedArray, ...leftArray, ...rightArray];
}

const testArray = [8, 20, -2, 4, -6];
console.log(mergeSort(testArray)); // [ -6, -2, 4, 8, 20 ]

// Big-O = O(n log n)

File: algorithms/sorting/python/merge_sort.py
def merge_sort(arr):
    # base case
    if len(arr) < 2:
        return arr
    mid = len(arr) // 2
    left_array = arr[:mid]
    right_array = arr[mid:]
    return merge(merge_sort(left_array), merge_sort(right_array))


def merge(left_array, right_array):
    sorted_array = []
    while left_array and right_array:
        if left_array[0] <= right_array[0]:
            # use pop since element should be removed from left array
            sorted_array.append(left_array.pop(0))
        else:
            sorted_array.append(right_array.pop(0))
    return sorted_array + left_array + right_array


test_array = [8, 20, -2, 4, -6]
print(merge_sort(test_array))  # [ -6, -2, 4, 8, 20 ]
